core/src/main/java/org/apache/druid/java/util/metrics/AllocationMetricCollector.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.metrics;
+
+import org.apache.druid.java.util.common.logger.Logger;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+class AllocationMetricCollector
+{
+  private static final Logger log = new Logger(AllocationMetricCollector.class);
+  private Method getThreadAllocatedBytes;
+  private ThreadMXBean threadMXBean;
+  private Map<Long, Long> previousResults;
+  private boolean initialized = false;
+
+  AllocationMetricCollector()
+  {
+    try {
+      // no Java official public API is present for the following method, so using reflection
+      threadMXBean = ManagementFactory.getThreadMXBean();
+      getThreadAllocatedBytes = threadMXBean.getClass().getMethod("getThreadAllocatedBytes", long[].class);
+      getThreadAllocatedBytes.setAccessible(true);
+      long[] allThreadIds = threadMXBean.getAllThreadIds();
+      long[] bytes = (long[]) getThreadAllocatedBytes.invoke(threadMXBean, (Object) allThreadIds);
+
+      previousResults = new HashMap<>(allThreadIds.length);
+      for (int i = 0; i < allThreadIds.length; i++) {
+        long threadId = allThreadIds[i];
+        previousResults.put(threadId, bytes[i]);
+
+      }
+      initialized = true;
+    }
+    catch (Exception e) {
```

Comment:
There is a whole lot of runtime exception types which can be thrown by reflection API (ie IllegalArgumentException and NPE)

Reply:
I don't see how IllegalArgumentException and NPE are possible. `NoSuchMethodException` should be rethrown, because it's a programming mistake. `IllegalAccessException` and `SecurityException` should be reported in logs.

Reply:
`NoSuchMethodException` can be due to a JVM private API change.

Reply:
You are right.